File: src/main/java/com/github/pozo/java/collections/map/Maps.java
package com.github.pozo.java.collections.map;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Maps {
    public static void main(String[] args) {
        lazyMap();
    }

    static void lazyMap() {
        Map<String, List<Integer>> map = new HashMap<>();
        final ArrayList<Integer> integers = new ArrayList<>();
        integers.add(32);
        map.put("aaa", integers);

        final String key = "aaa";
        final int value = 2;

        if (!map.containsKey(key)) {
            ArrayList<Integer> listOfValues = new ArrayList<>();
            listOfValues.add(value);
            map.put(key, listOfValues);
        } else {
            map.get(key).add(value);
        }
        System.out.println(map);
    }

}

File: src/main/kotlin/com/github/pozo/kotlin/collections/map/Maps.kt
package com.github.pozo.kotlin.collections.map

fun main(args: Array<String>) {
    lazyMap()
}

fun mapOf_withDefault_getOrElse() {
    val map = mapOf("aaa" to 32, "bbb" to 44)//.withDefault { "asd" to 2 }

    val value = map.getOrElse("ccc", foo("a"))

    println(value)
}

fun mapOf_withDefault_getOrElse_WithType() {
    val map = mapOf("aaa" to 32, "bbb" to 44)//.withDefault { "asd" to 2 }

    // invalid
    // val value: Int = map.getOrElse("ccc", foo("a"))
    val value: Int = map.getOrElse("ccc", { 2 })

    println(value)
}

fun foo(x: String): () -> String = { x + "!" }

fun mapOf_withDefault_OnEach() {
    val map = mapOf("aaa" to 32, "bbb" to 44).withDefault { "asd" to 2 }

    map.onEach { println(it) }
}

fun mapOf_withDefault() {
    val map = mapOf("aaa" to 32, "bbb" to 44).withDefault { "asd" to 2 }

    val value = map.getValue("a")

    println(value)
}

fun mapOf_withDefault_ExpectedAsd_ButReturnNull() {
    val map = mapOf("aaa" to 32, "bbb" to 44).withDefault { "asd" to 2 }

    val value = map["a"]

    println(value)
}

fun mapOf_getOrDefault() {
    val map = mapOf("aaa" to 32, "bbb" to 44)

    val value = map.getOrDefault("asd", 23)

    println(value)
}

fun mutableMapOf() {
    val map = mutableMapOf("aaa" to 32, "bbb" to 44)

    val value = map.getOrPut("ccc", { 23 })

    println(value)
}

fun lazyMap() {
    val map = mutableMapOf("aaa" to mutableListOf(32))

    map.putIfAbsent("aaa", mutableListOf(2))?.add(2)

    println(map)
}